src/SystemState.ts

```diff
@@ -101,4 +106,45 @@ export class SystemState {
         checkIntegerInRange("Group index", iGroup, 0, this._nGroups - 1);
         checkIntegerInRange("Particle index", iParticle, 0, this._nParticles - 1);
     }
+
+    /**
+     * Apply a reordering to the particles in every group according to a reordering parameter which defines
+     * a new order for each group.
+     * @param reordering {@link https://github.com/scijs/ndarray | NdArray } whose first dimension defines group index
+     * and whose second dimension defines a new order of particles for each group. For example, if this state has three
+     * particles per group, a reordering of `[2, 0, 1]` for any group would imply that its current third particle should
+     * be reordered to the first position, its current first particle to the second position, and its current second
+     * particle to the third position.
+     */
+    public reorder(reordering: ndarray.NdArray) {
+        const shape = reordering.shape;
+        if (shape.length !== 2 || shape[0] !== this._nGroups || shape[1] !== this._nParticles) {
+            throw new DustParameterError(
+                "Unexpected reordering shape. " +
+                    `Expected [${this._nGroups},${this._nParticles}] but got ${JSON.stringify(shape)}.`
+            );
+        }
+        // Check that each reordering contains expected values
+        for (let iGroup = 0; iGroup < this._nGroups; iGroup++) {
+            const reorder = particleStateToArray(reordering.pick(iGroup, null));
+            const sortedReorder = reorder.toSorted();
+            // Each reordering should contain every index value once so, when sorted, each value should equal its index
+            if (sortedReorder.some((value: number, index: number) => value !== index)) {
+                throw new DustParameterError(`Unexpected reorder values: ${JSON.stringify(reorder)}`);
+            }
+        }
+
+        const reordered = this.newState();
```

Comment:
This is fine for now, but we might refactor in future (though it would be worth getting some benchmarks sorted first - which suggests we might want the whole filter done).  In the C++ code the runner (the `DiscreteSystem` here) holds `state` and `stateNext`, and we flip between these during the model run.  Then in the reorder we reorder from `state` to `stateNext` and do a swap.  I have no idea how JS will handle this, and it's possible that it will be just fine - this detail is fairly opaque to anything using the system

Reply:
I could do that here as well actually.